packages/adminrouter/extra/src/lib/cache.lua

```diff
@@ -238,23 +235,26 @@ local function fetch_and_store_marathon_apps(auth_token)
        end
 
        local ports = task["ports"] --task host port mapping by default
-       if is_ip_per_task(app) then
+       -- task["ports"] will contain the host ports made available to the task:
+       -- see https://mesosphere.github.io/marathon/docs/ports.html#definitions,
+       -- note on the hostPort definitions.
+       -- AR, however, needs the container port to route the request to.
+       -- In "container" mode we find the container port out from portMappings array
+       -- for the case when container port is fixed (non-zero value specified).
+       -- When container port is specified as 0 it will be randomly assigned
+       -- by Marathon and the actual container port will be the same as the host port:
+       -- https://mesosphere.github.io/marathon/docs/ports.html#random-port-assignment
+       -- We do not override it with container port from the portMappings array
+       -- in that case.
+       -- In "container/bridge" and "host" networking modes we need to use the
+       -- host port for routing (available via task's ports array)
+       if is_container_network(app) and app["container"]["portMappings"][portIdx]["containerPort"] ~= 0 then
          ports = {}
-         if is_user_network(app) then
-            -- override with ports from the container's portMappings
-            local port_mappings = app["container"]["docker"]["portMappings"] or app["portDefinitions"] or {}
-            local port_attr = app["container"]["docker"]["portMappings"] and "containerPort" or "port"
-            for _, port_mapping in ipairs(port_mappings) do
-               table.insert(ports, port_mapping[port_attr])
-            end
-         else
-            --override with the discovery ports
-            local discovery_ports = app["ipAddress"]["discovery"]["ports"]
-            for _, discovery_port in ipairs(discovery_ports) do
-                table.insert(ports, discovery_port["number"])
-            end
-         end
-       end
+         local port_mappings = app["container"]["portMappings"] or {}
+          for _, port_mapping in ipairs(port_mappings) do
```

Comment:
This seems to be a lot of effort to get a single port. At this point we know that `app["container"]["portMappings"][portIdx]["containerPort"]` contains the value we want for `port`. Do we need to build a whole other array?

Could we just have something like:
```
if is_container_network(app) and app["container"]["portMappings"][portIdx]["containerPort"] ~= 0 then
    port = app["container"]["portMappings"][portIdx]["containerPort"]
else
    port = task["ports"][portIdx]
end
```